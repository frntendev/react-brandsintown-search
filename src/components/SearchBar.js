import React from "react";
import PropTypes from "prop-types";
import { css, cx } from "react-emotion";

const styles = fullMode => ({
  root: css`
    max-width: ${!fullMode && "510px"};
    margin: 160px auto;
    text-align: right;
    label: root;
    @media (max-width: 480px) {
      width: 90%;
      margin: 50px auto;
    }
  `,
  fullRoot: css`
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    background: #fff;
    z-index: 9;
    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
    transition: 0.3s all ease;
  `,
  textfieldArea: css`
    width: 100%;
    display: grid;
    display: -ms-grid;
    grid-template-rows: 1fr 50px;
    -ms-grid-rows: 1fr 5px 50px;
    grid-template-columns: 1fr;
    -ms-grid-columns: 1fr;
    grid-gap: 5px;
    text-align: center;
    justify-content: center;
    -ms-grid-row-align: center;
    -ms-grid-column-align: center;
    color: gray;
    label: textfieldArea;
    @media (max-width: 480px) {
      margin-top: 50px;
      grid-template-columns: 1fr;
    }
    &:after {
      position: relative;
      top: 0;
      z-index: 2;
      content: "";
      display: block;
      height: 2px;
      width: 0;
      background: transparent;
      transition: width 0.5s ease, background-color 0.5s ease;
    }
    &:hover:after {
      width: 100%;
      background: #000;
    }
  `,
  title: css`
    margin: 0;
    color: #3a3c30;
    font-size: 2.5rem;
    -ms-grid-row: 1;
    label: title;
    @media (max-width: 480px) {
      font-size: 1.8rem;
    }
  `,
  form: css`
    position: ${!fullMode && "relative"};
    -ms-grid-row: 3;
    label: form;
    &:after {
      width: 100%;
      position: relative;
      top: 5px;
      z-index: 2;
      content: "";
      display: block;
      margin: auto;
      height: 2px;
      background: #aaa;
    }
    ${fullMode && fullInput};
  `,
  transparent: css`
    background: transparent;
    border: none;
    color: inherit;
    outline: 0;
    label: transparent;
  `,
  button: css`
    font-size: 1.6em;
    label: button;
  `,
  input: css`
    -webkit-appearance: none;
    width: 100%;
    height: 100%;
    font-size: 1.6rem;
    padding-left: 45px;
    label: input;
  `,
  icon: css`
    width: 20px;
    height: 20px;
    position: absolute;
    left: ${fullInput ? "10px" : 0};
    top: 50%;
    transform: translateY(-50%);
    fill: #aaa;
    label: icon;
  `,
  clearHistory: css`
    background: transparent;
    color: #ef0909;
    font-weight: bold;
    text-transform: uppercase;
    font-size: 0.6rem;
    cursor: pointer;
    margin-top: 15px;
    border-radius: 2px;
    transition: 0.3s background ease;
    padding: 10px;
    border: none;
    outline: none;
    opacity: ${fullMode && 0};
    label: clearHistory;
    &:hover {
      background: #ef090938;
    }
  `
});

const fullInput = css`
  width: 100%;
  position: fixed;
  -ms-grid-row: 3;
  top: 0;
  left: 0;
  z-index: 9;
  background: #fff;
  height: 50px;
  &:after {
    top: 0;
  }
`;

class SearchBar extends React.Component {
  // Initial state that contains input value and fullMode
  // fullMode is an state that works after scrolling down
  state = { name: this.props.value || "", fullMode: false };

  // A function that keeps input value
  onChange = event => {
    this.setState({
      name: event.target.value
    });
  };

  // Attaching an event to `window.scroll`
  componentDidMount() {
    window.addEventListener("scroll", this.handleScroll);
  }

  // Check `window.scroll` and if more that 320 then shows a full-width
  // text-input at the top of the page
  handleScroll = () => {
    this.setState({
      fullMode: window.scrollY > 320
    });
  };

  // Pass input value to parent component
  handleSearch = event => {
    if (event) event.preventDefault();
    if (this.state.name !== "") this.props.onSubmit(this.state.name);
  };

  render() {
    const props = this.props;
    const style = styles(this.state.fullMode);
    return (
      <div className={style.root}>
        <div className={style.textfieldArea}>
          <h1 className={style.title}>Search for your favorite artist</h1>
          <form
            type="get"
            action="search"
            className={style.form}
            onSubmit={this.handleSearch}
          >
            <input
              id="artist-input"
              name="name"
              className={cx(style.input, style.transparent)}
              type="search"
              placeholder={props.placeholder}
              value={this.state.name}
              onChange={this.onChange}
              autoComplete="off"
            />
            <svg className={style.icon} viewBox="0 0 16 16">
              <path d="M15.56 15.56c-0.587 0.587-1.538 0.587-2.125 0l-2.652-2.652c-1.090 0.699-2.379 1.116-3.771 1.116-3.872 0-7.012-3.139-7.012-7.012s3.14-7.012 7.012-7.012c3.873 0 7.012 3.139 7.012 7.012 0 1.391-0.417 2.68-1.116 3.771l2.652 2.652c0.587 0.587 0.587 1.538 0 2.125zM7.012 2.003c-2.766 0-5.009 2.242-5.009 5.009s2.243 5.009 5.009 5.009c2.766 0 5.009-2.242 5.009-5.009s-2.242-5.009-5.009-5.009z" />
            </svg>
          </form>
        </div>
        {props.value && (
          <button
            id="clear-history"
            className={style.clearHistory}
            onClick={props.onClearHistory}
          >
            Clear History
          </button>
        )}
      </div>
    );
  }
}
export default SearchBar;

SearchBar.propTypes = {
  /**
   * Input placeholder
   */
  placeholder: PropTypes.string,
  /**
   * Input value
   * Can be either string or integer
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * A callback function to remove search history from browser
   */
  onClearHistory: PropTypes.func
};
