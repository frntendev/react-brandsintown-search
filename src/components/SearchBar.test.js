import React from "react";
import { shallow, mount, configure } from "enzyme";
import SearchBar from "./SearchBar";
import Adapter from "enzyme-adapter-react-16";
configure({ adapter: new Adapter() });

describe("SearchBar", () => {
  test("renders component itself correctly", () => {
    const renderedComponent = shallow(<SearchBar />);
    expect(renderedComponent).toMatchSnapshot();
  });

  test("input value is sync state", () => {
    const renderedComponent = mount(<SearchBar />);
    renderedComponent.setState({ name: "Maroon 5" });
    expect(renderedComponent.find("input").props().value).toEqual("Maroon 5");
  });

  test("clear history button exists", () => {
    const renderedComponent = mount(<SearchBar value="Maroon 5" />);
    expect(renderedComponent.find("#clear-history").exists()).toEqual(true);
  });

  test("clear history button is not exist", () => {
    const renderedComponent = mount(<SearchBar />);
    expect(renderedComponent.find("#clear-history").exists()).toEqual(false);
  });

  test("input onChange simulation", () => {
    const renderedComponent = mount(<SearchBar />);
    renderedComponent.setState({ name: "Maroon 5" });
    expect(renderedComponent.find("input").props().value).toEqual("Maroon 5");
    renderedComponent
      .find("input")
      .simulate("change", { target: { value: "Shakira" } });
    expect(renderedComponent.find("input").props().value).toEqual("Shakira");
  });
});
