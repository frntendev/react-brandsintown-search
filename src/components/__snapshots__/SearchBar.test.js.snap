// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar renders component itself correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchBar />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="css-1h39y3a-textfieldArea"
        >
          <h1
            className="css-cdfr3j-title"
          >
            Search for your favorite artist
          </h1>
          <form
            action="search"
            className="css-n3e1v5-form"
            onSubmit={[Function]}
            type="get"
          >
            <input
              autoComplete="off"
              className="css-1u1pxsd-input-transparent"
              id="artist-input"
              name="name"
              onChange={[Function]}
              type="search"
              value=""
            />
            <svg
              className="css-1uqe3un-icon"
              viewBox="0 0 16 16"
            >
              <path
                d="M15.56 15.56c-0.587 0.587-1.538 0.587-2.125 0l-2.652-2.652c-1.090 0.699-2.379 1.116-3.771 1.116-3.872 0-7.012-3.139-7.012-7.012s3.14-7.012 7.012-7.012c3.873 0 7.012 3.139 7.012 7.012 0 1.391-0.417 2.68-1.116 3.771l2.652 2.652c0.587 0.587 0.587 1.538 0 2.125zM7.012 2.003c-2.766 0-5.009 2.242-5.009 5.009s2.243 5.009 5.009 5.009c2.766 0 5.009-2.242 5.009-5.009s-2.242-5.009-5.009-5.009z"
              />
            </svg>
          </form>
        </div>,
        undefined,
      ],
      "className": "css-156hyeb-root",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1
              className="css-cdfr3j-title"
            >
              Search for your favorite artist
            </h1>,
            <form
              action="search"
              className="css-n3e1v5-form"
              onSubmit={[Function]}
              type="get"
            >
              <input
                autoComplete="off"
                className="css-1u1pxsd-input-transparent"
                id="artist-input"
                name="name"
                onChange={[Function]}
                type="search"
                value=""
              />
              <svg
                className="css-1uqe3un-icon"
                viewBox="0 0 16 16"
              >
                <path
                  d="M15.56 15.56c-0.587 0.587-1.538 0.587-2.125 0l-2.652-2.652c-1.090 0.699-2.379 1.116-3.771 1.116-3.872 0-7.012-3.139-7.012-7.012s3.14-7.012 7.012-7.012c3.873 0 7.012 3.139 7.012 7.012 0 1.391-0.417 2.68-1.116 3.771l2.652 2.652c0.587 0.587 0.587 1.538 0 2.125zM7.012 2.003c-2.766 0-5.009 2.242-5.009 5.009s2.243 5.009 5.009 5.009c2.766 0 5.009-2.242 5.009-5.009s-2.242-5.009-5.009-5.009z"
                />
              </svg>
            </form>,
          ],
          "className": "css-1h39y3a-textfieldArea",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Search for your favorite artist",
              "className": "css-cdfr3j-title",
            },
            "ref": null,
            "rendered": "Search for your favorite artist",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "action": "search",
              "children": Array [
                <input
                  autoComplete="off"
                  className="css-1u1pxsd-input-transparent"
                  id="artist-input"
                  name="name"
                  onChange={[Function]}
                  type="search"
                  value=""
                />,
                <svg
                  className="css-1uqe3un-icon"
                  viewBox="0 0 16 16"
                >
                  <path
                    d="M15.56 15.56c-0.587 0.587-1.538 0.587-2.125 0l-2.652-2.652c-1.090 0.699-2.379 1.116-3.771 1.116-3.872 0-7.012-3.139-7.012-7.012s3.14-7.012 7.012-7.012c3.873 0 7.012 3.139 7.012 7.012 0 1.391-0.417 2.68-1.116 3.771l2.652 2.652c0.587 0.587 0.587 1.538 0 2.125zM7.012 2.003c-2.766 0-5.009 2.242-5.009 5.009s2.243 5.009 5.009 5.009c2.766 0 5.009-2.242 5.009-5.009s-2.242-5.009-5.009-5.009z"
                  />
                </svg>,
              ],
              "className": "css-n3e1v5-form",
              "onSubmit": [Function],
              "type": "get",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "autoComplete": "off",
                  "className": "css-1u1pxsd-input-transparent",
                  "id": "artist-input",
                  "name": "name",
                  "onChange": [Function],
                  "placeholder": undefined,
                  "type": "search",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <path
                    d="M15.56 15.56c-0.587 0.587-1.538 0.587-2.125 0l-2.652-2.652c-1.090 0.699-2.379 1.116-3.771 1.116-3.872 0-7.012-3.139-7.012-7.012s3.14-7.012 7.012-7.012c3.873 0 7.012 3.139 7.012 7.012 0 1.391-0.417 2.68-1.116 3.771l2.652 2.652c0.587 0.587 0.587 1.538 0 2.125zM7.012 2.003c-2.766 0-5.009 2.242-5.009 5.009s2.243 5.009 5.009 5.009c2.766 0 5.009-2.242 5.009-5.009s-2.242-5.009-5.009-5.009z"
                  />,
                  "className": "css-1uqe3un-icon",
                  "viewBox": "0 0 16 16",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "d": "M15.56 15.56c-0.587 0.587-1.538 0.587-2.125 0l-2.652-2.652c-1.090 0.699-2.379 1.116-3.771 1.116-3.872 0-7.012-3.139-7.012-7.012s3.14-7.012 7.012-7.012c3.873 0 7.012 3.139 7.012 7.012 0 1.391-0.417 2.68-1.116 3.771l2.652 2.652c0.587 0.587 0.587 1.538 0 2.125zM7.012 2.003c-2.766 0-5.009 2.242-5.009 5.009s2.243 5.009 5.009 5.009c2.766 0 5.009-2.242 5.009-5.009s-2.242-5.009-5.009-5.009z",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "path",
                },
                "type": "svg",
              },
            ],
            "type": "form",
          },
        ],
        "type": "div",
      },
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="css-1h39y3a-textfieldArea"
          >
            <h1
              className="css-cdfr3j-title"
            >
              Search for your favorite artist
            </h1>
            <form
              action="search"
              className="css-n3e1v5-form"
              onSubmit={[Function]}
              type="get"
            >
              <input
                autoComplete="off"
                className="css-1u1pxsd-input-transparent"
                id="artist-input"
                name="name"
                onChange={[Function]}
                type="search"
                value=""
              />
              <svg
                className="css-1uqe3un-icon"
                viewBox="0 0 16 16"
              >
                <path
                  d="M15.56 15.56c-0.587 0.587-1.538 0.587-2.125 0l-2.652-2.652c-1.090 0.699-2.379 1.116-3.771 1.116-3.872 0-7.012-3.139-7.012-7.012s3.14-7.012 7.012-7.012c3.873 0 7.012 3.139 7.012 7.012 0 1.391-0.417 2.68-1.116 3.771l2.652 2.652c0.587 0.587 0.587 1.538 0 2.125zM7.012 2.003c-2.766 0-5.009 2.242-5.009 5.009s2.243 5.009 5.009 5.009c2.766 0 5.009-2.242 5.009-5.009s-2.242-5.009-5.009-5.009z"
                />
              </svg>
            </form>
          </div>,
          undefined,
        ],
        "className": "css-156hyeb-root",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1
                className="css-cdfr3j-title"
              >
                Search for your favorite artist
              </h1>,
              <form
                action="search"
                className="css-n3e1v5-form"
                onSubmit={[Function]}
                type="get"
              >
                <input
                  autoComplete="off"
                  className="css-1u1pxsd-input-transparent"
                  id="artist-input"
                  name="name"
                  onChange={[Function]}
                  type="search"
                  value=""
                />
                <svg
                  className="css-1uqe3un-icon"
                  viewBox="0 0 16 16"
                >
                  <path
                    d="M15.56 15.56c-0.587 0.587-1.538 0.587-2.125 0l-2.652-2.652c-1.090 0.699-2.379 1.116-3.771 1.116-3.872 0-7.012-3.139-7.012-7.012s3.14-7.012 7.012-7.012c3.873 0 7.012 3.139 7.012 7.012 0 1.391-0.417 2.68-1.116 3.771l2.652 2.652c0.587 0.587 0.587 1.538 0 2.125zM7.012 2.003c-2.766 0-5.009 2.242-5.009 5.009s2.243 5.009 5.009 5.009c2.766 0 5.009-2.242 5.009-5.009s-2.242-5.009-5.009-5.009z"
                  />
                </svg>
              </form>,
            ],
            "className": "css-1h39y3a-textfieldArea",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Search for your favorite artist",
                "className": "css-cdfr3j-title",
              },
              "ref": null,
              "rendered": "Search for your favorite artist",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "action": "search",
                "children": Array [
                  <input
                    autoComplete="off"
                    className="css-1u1pxsd-input-transparent"
                    id="artist-input"
                    name="name"
                    onChange={[Function]}
                    type="search"
                    value=""
                  />,
                  <svg
                    className="css-1uqe3un-icon"
                    viewBox="0 0 16 16"
                  >
                    <path
                      d="M15.56 15.56c-0.587 0.587-1.538 0.587-2.125 0l-2.652-2.652c-1.090 0.699-2.379 1.116-3.771 1.116-3.872 0-7.012-3.139-7.012-7.012s3.14-7.012 7.012-7.012c3.873 0 7.012 3.139 7.012 7.012 0 1.391-0.417 2.68-1.116 3.771l2.652 2.652c0.587 0.587 0.587 1.538 0 2.125zM7.012 2.003c-2.766 0-5.009 2.242-5.009 5.009s2.243 5.009 5.009 5.009c2.766 0 5.009-2.242 5.009-5.009s-2.242-5.009-5.009-5.009z"
                    />
                  </svg>,
                ],
                "className": "css-n3e1v5-form",
                "onSubmit": [Function],
                "type": "get",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "autoComplete": "off",
                    "className": "css-1u1pxsd-input-transparent",
                    "id": "artist-input",
                    "name": "name",
                    "onChange": [Function],
                    "placeholder": undefined,
                    "type": "search",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <path
                      d="M15.56 15.56c-0.587 0.587-1.538 0.587-2.125 0l-2.652-2.652c-1.090 0.699-2.379 1.116-3.771 1.116-3.872 0-7.012-3.139-7.012-7.012s3.14-7.012 7.012-7.012c3.873 0 7.012 3.139 7.012 7.012 0 1.391-0.417 2.68-1.116 3.771l2.652 2.652c0.587 0.587 0.587 1.538 0 2.125zM7.012 2.003c-2.766 0-5.009 2.242-5.009 5.009s2.243 5.009 5.009 5.009c2.766 0 5.009-2.242 5.009-5.009s-2.242-5.009-5.009-5.009z"
                    />,
                    "className": "css-1uqe3un-icon",
                    "viewBox": "0 0 16 16",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "d": "M15.56 15.56c-0.587 0.587-1.538 0.587-2.125 0l-2.652-2.652c-1.090 0.699-2.379 1.116-3.771 1.116-3.872 0-7.012-3.139-7.012-7.012s3.14-7.012 7.012-7.012c3.873 0 7.012 3.139 7.012 7.012 0 1.391-0.417 2.68-1.116 3.771l2.652 2.652c0.587 0.587 0.587 1.538 0 2.125zM7.012 2.003c-2.766 0-5.009 2.242-5.009 5.009s2.243 5.009 5.009 5.009c2.766 0 5.009-2.242 5.009-5.009s-2.242-5.009-5.009-5.009z",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "path",
                  },
                  "type": "svg",
                },
              ],
              "type": "form",
            },
          ],
          "type": "div",
        },
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
